presto-orc/src/main/java/com/facebook/presto/orc/writer/StreamOrderingLayout.java

```diff
@@ -0,0 +1,219 @@
+        private final Map<ColIdKeyId, Integer> keyIdToSeqId;
+
+        public StreamMetadata(Map<ColIdSequenceId, List<StreamDataOutput>> seqIdToStreams, Map<ColIdKeyId, Integer> keyIdToSeqId)
+        {
+            this.seqIdToStreams = requireNonNull(seqIdToStreams, "seqIdToStreams cannot be null");
+            this.keyIdToSeqId = requireNonNull(keyIdToSeqId, "keyIdToSeqId cannot be null");
+        }
+    }
+
+    private static class ColIdKeyId
+    {
+        private final int columnId;
+        private final DwrfProto.KeyInfo keyInfo;
+
+        public ColIdKeyId(int inputColumnId, DwrfProto.KeyInfo inputKeyId)
+        {
+            this.columnId = inputColumnId;
+            this.keyInfo = requireNonNull(inputKeyId, "inputKeyId cannot be null");
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            if (obj == null) {
+                return false;
+            }
+            if (!(obj instanceof ColIdKeyId)) {
+                return false;
+            }
+            ColIdKeyId input = (ColIdKeyId) obj;
+            return this.columnId == input.columnId && this.keyInfo.equals(input.keyInfo);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(columnId, keyInfo);
+        }
+    }
+
+    private static class ColIdSequenceId
+    {
+        private final int columnId;
+        private final int sequenceId;
+
+        public ColIdSequenceId(int inputColumnId, int inputSequenceId)
+        {
+            this.columnId = inputColumnId;
+            this.sequenceId = inputSequenceId;
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            if (obj == null) {
+                return false;
+            }
+            if (!(obj instanceof ColIdSequenceId)) {
+                return false;
+            }
+            ColIdSequenceId input = (ColIdSequenceId) obj;
+            return this.columnId == input.columnId && this.sequenceId == input.sequenceId;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(columnId, sequenceId);
+        }
+    }
+
+    private StreamMetadata getStreamMetadata(
+            Map<Integer, Integer> nodeIdToColumnId,
+            Map<Integer, ColumnEncoding> nodeIdToColumnEncodings,
+            DwrfStreamOrderingConfig config)
+    {
+        ImmutableMap.Builder<ColIdKeyId, Integer> keyIdToSeqIdBuilder = ImmutableMap.<ColIdKeyId, Integer>builder();
+        ImmutableMap.Builder<ColIdSequenceId, List<StreamDataOutput>> seqIdToStreamBuilder = ImmutableMap.<ColIdSequenceId, List<StreamDataOutput>>builder();
+        Map<Integer, Set<DwrfProto.KeyInfo>> colIdToKeyIds = new HashMap<>();
+
+        for (Map.Entry<Integer, List<DwrfProto.KeyInfo>> orderingConfig : config.getStreamOrdering().entrySet()) {
+            colIdToKeyIds.put(orderingConfig.getKey(), new HashSet<>(orderingConfig.getValue()));
```

Comment:
Should we do it right away in the config? Converting a long list to a set for every stripe seems expensive.

Reply:
Yes, I think this can be cached inside the config